File: backend/pkg/controllers/service.go
package controllers

import (
	"fmt"
	"log"
	"net/http"

	"area/pkg/models"

	"github.com/gin-gonic/gin"
)

// CreateServiceInput struct
type CreateServiceInput struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description" binding:"required"`
}

// CreateService function
func CreateService(c *gin.Context) {
	var input CreateServiceInput
	if err := c.ShouldBindJSON(&input); err != nil {
		fmt.Println("error is : ", err)
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	service := models.Service{Name: input.Name, Description: input.Description}
	models.DB.Create(&service)
	c.JSON(http.StatusOK, gin.H{"data": service})
}

func GetService(c *gin.Context) {
	var service models.Service
	id := c.Param("id")

	if result := models.DB.First(&service, id); result.Error != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Record not found!"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": service})
}

// GetServices func
func GetServices(c *gin.Context) {
	var services []models.Service

	if result := models.DB.Find(&services); result.Error != nil {
		panic(result.Error)
	}
	c.JSON(http.StatusOK, gin.H{"data": services})
}

// UpdateServiceInput struct
type UpdateServiceInput struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description" binding:"required"`
}

// UpdateService function
func UpdateService(c *gin.Context) {
	var service models.Service
	id := c.Param("id")

	log.Println("UpdateService request received")
	if result := models.DB.First(&service, id); result.Error != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Record not found!"})
		return
	}

	var input UpdateServiceInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updateService := models.Service{Name: input.Name, Description: input.Description}
	models.DB.Model(&service).Updates(updateService)

	c.JSON(http.StatusOK, gin.H{"data": service})
}

// DeleteService func
func DeleteService(c *gin.Context) {
	var service models.Service
	id := c.Param("id")

	log.Println("DeleteService request received")
	if result := models.DB.First(&service, id); result.Error != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Record not found!"})
		return
	}

	models.DB.Delete(&service)
	c.JSON(http.StatusOK, gin.H{"data": true})
}

File: frontend/src/store/service.js
import { defineStore } from "pinia";
import axios from "axios";

export const useServiceStore = defineStore("service", {
    state: () => ({
        services: [],
        service: {},
        loading: false,
        error: null,
    }),
    getters: {
        getServices: (state) => state.services,
        getService: (state) => state.service,
        getLoading: (state) => state.loading,
        getError: (state) => state.error,
    },
    actions: {
        async getAllServices() {
            this.loading = true;
            axios.get("services")
            .then((response) => {
                this.services = response.data.data;
            })
            .catch((error) => {
                console.log(error);
                this.error = error;
            });
            this.loading = false;
        },
        async getService(id) {
            this.loading = true;
            axios.get(`services/${id}`)
            .then((response) => {
                this.service = response.data.data;
            })
            .catch((error) => {
                console.log(error);
                this.error = error;
            });
            this.loading = false;
        },
        async addService(name, description) {
            this.loading = true;
            axios.post("services", {
                name,
                description,
            })
            .then((response) => {
                this.services.push(response.data.data);
                
            })
            .catch((error) => {
                console.log(error);
                this.error = error;
            });
            this.loading = false;
        },
        async updateService(id, service) {
            this.loading = true;
            axios.put(`services/${id}`, service)
            .then((response) => {
                this.services = this.services.map((service) => {
                    if (service.id === id) {
                        return response.data.data;
                    }
                    return service;
                });
            })
            .catch((error) => {
                console.log(error);
                this.error = error;
            });
            this.loading = false;
        },
        async deleteService(id) {
            this.loading = true;
            axios.delete(`services/${id}`)
            .then((response) => {
                console.log(response.data.data);
                this.services = this.services.filter((service) => service.ID !== id);
            })
            .catch((error) => {
                console.log(error);
                this.error = error;
            })
            this.loading = false;
        },
    },
});

    